bookwyrm/preview_images.py

```diff
@@ -0,0 +1,367 @@
+        )
+    else:
+        image_bg_color = BG_COLOR
+
+    # Background (using the color)
+    img = Image.new("RGBA", (IMG_WIDTH, IMG_HEIGHT), color=image_bg_color)
+
+    # Contents
+    inner_img_x = margin + inner_img_width - inner_img_layer.width
+    inner_img_y = math.floor((IMG_HEIGHT - inner_img_layer.height) / 2)
+    content_x = margin + inner_img_width + gutter
+    content_width = IMG_WIDTH - content_x - margin
+
+    contents_layer = Image.new(
+        "RGBA", (content_width, IMG_HEIGHT), color=TRANSPARENT_COLOR
+    )
+    contents_composite_y = 0
+
+    if show_instance_layer:
+        instance_layer = generate_instance_layer(content_width)
+        contents_layer.alpha_composite(instance_layer, (0, contents_composite_y))
+        contents_composite_y = contents_composite_y + instance_layer.height + gutter
+
+    texts_layer = generate_texts_layer(texts, content_width)
+    contents_layer.alpha_composite(texts_layer, (0, contents_composite_y))
+    contents_composite_y = contents_composite_y + texts_layer.height + gutter
+
+    if rating:
+        # Add some more margin
+        contents_composite_y = contents_composite_y + gutter
+        rating_layer = generate_rating_layer(rating, content_width)
+        contents_layer.alpha_composite(rating_layer, (0, contents_composite_y))
+        contents_composite_y = contents_composite_y + rating_layer.height + gutter
+
+    contents_layer_box = contents_layer.getbbox()
+    contents_layer_height = contents_layer_box[3] - contents_layer_box[1]
+
+    contents_y = math.floor((IMG_HEIGHT - contents_layer_height) / 2)
+
+    if show_instance_layer:
+        # Remove Instance Layer from centering calculations
+        contents_y = contents_y - math.floor((instance_layer.height + gutter) / 2)
+
+    if contents_y < margin:
+        contents_y = margin
+
+    # Composite layers
+    img.paste(
+        inner_img_layer, (inner_img_x, inner_img_y), inner_img_layer.convert("RGBA")
+    )
+    img.alpha_composite(contents_layer, (content_x, contents_y))
+
+    return img
+
+
+def save_and_cleanup(image, instance=None):
+    if instance:
+        file_name = "%s.png" % str(uuid4())
+        image_buffer = BytesIO()
+
+        try:
+            try:
+                old_path = instance.preview_image.path
+            except ValueError:
+                old_path = ""
+
+            # Save
+            image.save(image_buffer, format="png")
+            instance.preview_image = InMemoryUploadedFile(
+                ContentFile(image_buffer.getvalue()),
+                "preview_image",
+                file_name,
+                "image/png",
+                image_buffer.tell(),
+                None,
+            )
+            instance.save(update_fields=["preview_image"])
+
+            # Clean up old file after saving
+            if os.path.exists(old_path):
+                os.remove(old_path)
+        finally:
+            image_buffer.close()
+
+
+@app.task
```

Comment:
Is it wise to have three different tasks? Or should we have only one task, accepting two arguments, `type` (`edition`, `user`, `site`) and `id` (`book.id`, `user.id`), calling each of these 3 methods depending on `type`.

Reply:
I think it makes sense for them to remain as separate tasks, it seems clear to me and I don't think it makes a difference for performance